pypdf/_writer.py

```diff
@@ -1679,75 +1679,214 @@ def compress_identical_objects(
         remove_orphans: bool = True,
     ) -> None:
         """
-        Parse the PDF file and merge objects that have the same hash.
-        This will make objects common to multiple pages.
-        Recommended to be used just before writing output.
+        Compress identical objects in the PDF file.
 
         Args:
             remove_identicals: Remove identical objects.
             remove_orphans: Remove unreferenced objects.
+        """
+        if not remove_identicals and not remove_orphans:
+            logger_warning(
+                "Remove identical objects and remove unreferenced objects are both disabled. Nothing has been done.",
+                __name__,
+            )
+            return
+
+        orphans: Optional[list[bool]] = None
 
+        if remove_identicals:
+            # When remove_orphans=True, also mark reachability while rewriting refs
+            orphans = self._remove_identical_objects(mark_orphans=remove_orphans)
+
+        if remove_orphans:
+            if orphans is None:
+                # recursive DFS (mark phase)
+                orphans = self._perform_orphan_recursive_mark_phase()
+            # sweep phase
+            assert orphans is not None
+            self._perform_orphan_sweep_phase(orphans)
+
+
+    def _remove_identical_objects(self, mark_orphans: bool) -> Optional[list[bool]]:
         """
+        Detect identical objects by hash and collapse duplicates.
+        If mark_orphans is True, mark reachable objects during the rewrite so that the caller
+        can later sweep truly unreachable ones.
+
+        This function:
+        - groups objects by obj.hash_value(),  ---  WARNING: this can cause collisions !!!
+        - keeps the first occurrence as the representative,
+        - for each duplicate sets self._objects[idx] = None (we do not insert an IndirectObject here),
+        - updates references inside *live* objects so that any IndirectObject that previously
+        pointed to a duplicate will point to the representative.
```

Comment:
I guess this should be indented?

```suggestion
          pointed to a duplicate will point to the representative.
```